### PR TITLE
Allow only single infix as in except branches. Fixes #7115

### LINKED ISSUE #7115: Compiler crash: multiple exception types with infix as
The following is allowed by semantic pass, but codegen crashes afterwards:

```
try:
  echo 10
except [ValueError as ex1, KeyError as ex2]:
  echo ex1.msg
```
Should it be rejected in semantic pass?

I can make a patch once I know what is the correct expected behavior

### PR DESCRIPTION
**Pull request by cooldome**
Rewrite ``semTry`` such that only single `as` infix expression is allowed.

I also changed commonType deduction to follow latest changes in ``semIf``, ``semCase`` for consistency

### CODE REVIEW

**Comment by cooldome** [comment]
Ready for review now

**Comment by dom96** [comment]
is it possible to write?

```nim
try: 
   echo 1
except KeyError as ex1, ValueError as ex2:
  echo 2
```

**Comment by cooldome** [comment]
@dom96 
No, it is possible. Only single `Exception as ex` is allowed in the except branch.
Previously, `semTry` was not checking that  `Exception as ex` is alone and allowed for that and compiler crashed in the later stages. Now `semTry` is checking it explicitly, no crash just proper error message. 
Next step, I plan is to use `Exception as ex` syntax to catch imported cpp exceptions.

**Comment by dom96** [comment]
It would be nice to support that syntax IMO.

**Comment by cooldome** [comment]
@dom96 
It quite problematic, what the following should mean:
```
try: 
   echo 1
except KeyError as ex1, ValueError as ex2:
  echo ex1.msg, "--", ex2.msg
```
One of `ex1` or `ex2` is always invalid expression as in the end only one exception will be handled at a time.

I was thinking if it is possible to support:

```
try: 
   echo 1
except (KeyError, ValueError) as ex:
  echo ex.msg
```
But it is also problematic, `ex `has no concrete type hence semantic check of except body will not pass.
I see only one realistic option is to support just a single `Exception as ex`

**Comment by dom96** [comment]
hrm, good point.

File: compiler/semstmts.nim [hunk 1]
```diff
@@ -252,69 +252,61 @@ proc semCase(c: PContext, n: PNode): PNode =
...
-    a.sons[length-1] = semExprBranchScope(c, a.sons[length-1])
-    if a.kind != nkFinally: typ = commonType(typ, a.sons[length-1].typ)
+    a[^1] = semExprBranchScope(c, a[^1])
+    if a.kind != nkFinally: typ = commonType(typ, a[^1])
     else: dec last
     closeScope(c)
 
   dec c.p.inTryStmt
-  if isEmptyType(typ) or typ.kind == tyNil:
+  if isEmptyType(typ) or typ.kind in {tyNil, tyExpr}:
```
Comment (Araq) [inline compiler/semstmts.nim]: Why is ``tyExpr`` added here?

Comment (cooldome) [reply compiler/semstmts.nim]: I have changed `` typ = commonType(typ, a[^1].typ)`` to ``typ = commonType(typ, a[^1])`` which makes try expression skip no return branches just like I did for if/case expressions in December. However, if all branches are no returning common type will be ``tyExpr``. Previously it was not possible.

Let me know if it is not desirable to skip noreturning branches in try expr.

File: compiler/semstmts.nim [hunk 2]
```diff
@@ -252,69 +252,61 @@ proc semCase(c: PContext, n: PNode): PNode =
     result.typ = typ
 
 proc semTry(c: PContext, n: PNode): PNode =
+
+  var check = initIntSet()
+  proc semExceptBranchType(typeNode: PNode): PNode =
+    let typ = semTypeNode(c, typeNode, nil).toObject()
+    if typ.kind != tyObject:
+      localError(typeNode.info, errExprCannotBeRaised)
+    if containsOrIncl(check, typ.id):
```
Comment (Araq) [inline compiler/semstmts.nim]: Pass ``check`` explicitly to ``semExceptBranchType`` in order to avoid the closure creation.

Comment (cooldome) [reply compiler/semstmts.nim]: Closure creation removed by using template instead

**Comment by cooldome** [comment]
I am somehow completely missed the fact that it was reviewed.
Comments are now resolved.

**Pull request merged**